ARCore/app/src/main/java/com/novoda/spikes/arcore/MainActivity.kt

```diff
@@ -0,0 +1,87 @@
+package com.novoda.spikes.arcore
+
+import android.opengl.GLSurfaceView
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.widget.Toast
+import com.google.ar.core.Session
+import com.google.ar.core.exceptions.CameraNotAvailableException
+import com.novoda.spikes.arcore.google.helper.TapHelper
+import com.novoda.spikes.arcore.helper.ARCoreDependenciesHelper
+import com.novoda.spikes.arcore.helper.CameraPermissionHelper
+import com.novoda.spikes.arcore.rendering.NovodaSurfaceViewRenderer
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var arSession: Session
+    private lateinit var renderer: NovodaSurfaceViewRenderer
+    private lateinit var debugViewDisplayer: DebugViewDisplayer
+    private lateinit var tapHelper: TapHelper
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        setupSurfaceView()
+    }
+
+    private fun setupSurfaceView() {
+        tapHelper = TapHelper(this)
+        debugViewDisplayer = DebugViewDisplayer(debugTextView)
+        renderer = NovodaSurfaceViewRenderer(this, debugViewDisplayer, tapHelper)
+
+        surfaceView.preserveEGLContextOnPause = true
+        surfaceView.setEGLContextClientVersion(2)
+        surfaceView.setEGLConfigChooser(8, 8, 8, 8, 16, 0) // Alpha used for plane blending.
+        surfaceView.setRenderer(renderer)
+        surfaceView.renderMode = GLSurfaceView.RENDERMODE_CONTINUOUSLY
+        surfaceView.setOnTouchListener(tapHelper)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        checkAREnvironment()
+    }
+
+    private fun checkAREnvironment() {
+        ARCoreDependenciesHelper.isARCoreIsInstalled(this).apply {
+            if (isARCoreInstalled && CameraPermissionHelper.isCameraPermissionGranted(this@MainActivity)) {
+                createOrResumeARSession()
+            } else {
+                showMessage(message)
+            }
+        }
+    }
+
+    private fun createOrResumeARSession() {
+        if (this::arSession.isInitialized.not()) {
```

Comment:
IMO this is a bit of a smell that the `arSession` should be a nullable `var` rather than a `lateinit var`, or maybe you want to use the `by lazy` initialiser if this just needs to happen once

Reply:
Updated here 01832ffb